src/snmalloc/mem/freelist.h

```diff
@@ -697,6 +687,16 @@ namespace snmalloc
           {
             length[i] = 0;
           }
+
+          // Head is not live when a building is initialised.
+          // We use this slot to store a pointer into the slab for the
+          // allocations. This then establishes the invariant that head is
+          // always (a possibly encoded) pointer into the slab, and thus
+          // the Freelist builder always knows which block it is referring too.
+          head[i] = Object::code_next(
+            address_cast(&head[i]),
+            reinterpret_cast<Object::T<BQueue>*>(slab),
```

Comment:
This might also be pronounced
```suggestion
            unsafe_from_uintptr<Object::T<BQueue>>(slab),
```
(Or maybe a `static_cast<uintptr_t>(slab)` is necessary.)  Whether that is more or less terrifying than `reinterpret_cast`, I will leave to you to decide.  I wonder if we want to wrap this up in something that makes it more obvious that it's just an address getting made into a provenance-free (and so, on CHERI, untagged) pointer?  `template<typename T> T* useless_ptr_from_addr(address_t)`?

Reply:
I tried this, but CHERI purecap build complained that I was creating a provenance free pointer.  Which is what I wanted to do :(